pkg/pool-weighted/contracts/managed/ManagedPoolSettings.sol

```diff
@@ -932,6 +947,110 @@ abstract contract ManagedPoolSettings is BasePool, ProtocolFeeCache, ReentrancyG
+     * If there is an ongoing gradual weight update, they will shift up or down with the current weight.
+     */
+    function getCurrentCircuitBreakerBounds(IERC20 token) public view returns (uint256, uint256) {
+        return
+            CircuitBreakerLib.getCurrentCircuitBreakerBounds(
+                _circuitBreakerState[token],
+                _getNormalizedWeight(token).complement()
+            );
+    }
+
+    /**
+     * @notice Returns flags indicating whether the lower or upper bound circuit breakers have tripped.
+     */
+    function hasCircuitBreakerTripped(IERC20 token) external view returns (bool, bool) {
+        return
+            CircuitBreakerLib.hasCircuitBreakerTripped(
+                _circuitBreakerState[token],
+                getActualSupply(),
+                _getNormalizedWeight(token),
+                _getUpscaledTokenBalance(token)
+            );
+    }
+
+    /**
+     * @notice Set a circuit breaker for a token.
+     * @dev This is a permissioned function, and disabled if the pool is paused. The lower and upper bounds
+     * are percentages, corresponding to a *relative* change in the token's spot price: e.g., a lower bound
+     * of 0.8 means the breaker should prevent trades that result in the value of the token dropping 20% or
+     * more relative to the rest of the pool.
+     */
+    function setCircuitBreaker(
+        IERC20 token,
+        uint256 lowerBoundPercentage,
+        uint256 upperBoundPercentage
+    ) external authenticate whenNotPaused {
+        _setCircuitBreaker(token, lowerBoundPercentage, upperBoundPercentage);
+    }
+
+    // Compute the reference values, then pass them along with the bounds to the library.
+    function _setCircuitBreaker(
+        IERC20 token,
+        uint256 lowerBoundPercentage,
+        uint256 upperBoundPercentage
+    ) private {
+        uint256 normalizedWeight = _getNormalizedWeight(token);
+
+        // Note that `getBptPrice` will revert if the token is invalid.
+        // The library will validate the lower/upper bounds
+        _circuitBreakerState[token] = CircuitBreakerLib.setCircuitBreakerFields(
+            getBptPrice(token, normalizedWeight),
+            normalizedWeight.complement(),
+            lowerBoundPercentage,
+            upperBoundPercentage
+        );
+
+        emit CircuitBreakerSet(token, lowerBoundPercentage, upperBoundPercentage);
+    }
+
+    /**
+     * @notice Get the price of a single token in terms of BPT, given the weight.
+     * @dev Returns an 18-decimal floating point number. Could be internal, but might be generally useful
+     * externally.
+     */
+    function getBptPrice(IERC20 token, uint256 normalizedWeight) public view returns (uint256) {
+        if (token == IERC20(this)) {
+            // BPT price of the BPT token itself is undefined.
+            _revert(Errors.INVALID_TOKEN);
```

Comment:
```suggestion
            return FixedPoint.ONE;
```
I don't think we need to revert here as the price of BPT in terms of BPT is defined.

Reply:
I went back and forth on that. I had it just returning ONE at first, then thought if somebody's doing this it's probably a mistake. But I suppose with highly composed pools it might get called this way.